### PR TITLE
pbp data with on court players

### PR DESCRIPTION
**Pull request by YigitAsik**
Added a function to pull play by play data and add on court players to each play.

### CODE REVIEW

File: .DS_Store
Comment (giasemidis) [inline .DS_Store:1]: please remove this file.

File: src/.DS_Store
Comment (giasemidis) [inline src/.DS_Store:1]: please remove this file

File: src/euroleague_api/team_lineup_euroleague_api.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+import pandas as pd
+import numpy as np
+from euroleague_api.play_by_play_data import PlayByPlay
+from euroleague_api.boxscore_data import BoxScoreData
+
+season = 2023
```
Comment (giasemidis) [inline src/euroleague_api/team_lineup_euroleague_api.py]: Please remove lines 6-13, they are not required for the function definition.

File: src/euroleague_api/team_lineup_euroleague_api.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+pd.set_option('display.width', 170)
+pd.set_option('display.max_rows', None)
+pd.set_option('display.float_format', lambda x: '%.3f' % x)
+
+def pbp_data_with_lineups(season, game_code):
+    
+    # Getting pbp for the match
+    pbp = PlayByPlay()
+    pbp_data = pbp.get_game_play_by_play_data(season=season, gamecode=game_code)
+    pbp_data['Indices'] = range(0, len(pbp_data)) # resetting indices since going to loop through them, don't want any duplicates
```
Comment (giasemidis) [inline src/euroleague_api/team_lineup_euroleague_api.py]: To reset indices, you can do `pbp_data = pbp_data.reset_index(drop=True)`

File: src/euroleague_api/team_lineup_euroleague_api.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    starting_five['ID'] = starting_five.groupby('Team').cumcount()
+    starting_five = starting_five.pivot(index='ID', columns='Team', values='Player')
+    starting_five.reset_index(drop=True, inplace=True)
+
+    # Creating player columns
+    on_court_cols = ['Team_A1', 'Team_A2', 'Team_A3', 'Team_A4', 'Team_A5', 'Team_B1', 'Team_B2', 'Team_B3', 'Team_B4', 'Team_B5']
+    pbp_data[on_court_cols] = np.nan
+
+    # Filling start of the match with starters
+    on_court_mask1 = ['Team_A1', 'Team_A2', 'Team_A3', 'Team_A4', 'Team_A5']
```
Comment (giasemidis) [inline src/euroleague_api/team_lineup_euroleague_api.py:23]: Advice, move `on_court_mask1` and `on_court_mask2` above `on_court_cols`, the do: `on_court_cols = on_court_cols1 + on_court_cols2`, because if you change an element in one of the variables `on_court_cols` should be in sync.

File: src/euroleague_api/team_lineup_euroleague_api.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    on_court_mask2 = ['Team_B1', 'Team_B2', 'Team_B3', 'Team_B4', 'Team_B5']
+
+    pbp_data.loc[0, on_court_mask1] = starting_five.iloc[:, 0].tolist()
+    pbp_data.loc[0, on_court_mask2] = starting_five.iloc[:, 1].tolist()
+
+    # Capturing subs and replacing player columns with players who substitute in
+    players = pbp_data.loc[min(pbp_data.index.tolist()), on_court_cols]
+
+    for idx in pbp_data.index.tolist(): # looping through plays of a game
```
Comment (giasemidis) [inline src/euroleague_api/team_lineup_euroleague_api.py]: To iterate the rows of a dataframe a cleaner solution is the following:

```
for idx, row in pbp_data.iterrows():
    if row['PLAYTYPE'] == 'IN':
    sub_time = row['MARKERTIME']
   ...
```
the code is cleaner

File: src/euroleague_api/team_lineup_euroleague_api.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                out_players_1 = []
+                for i in subs_team1.index: # capturing every sub made by the first team at that particular time in the match
+                    if subs_team1.loc[i, 'PLAYTYPE'] == 'IN':
+                        tmp_in_player = subs_team1.loc[i, 'PLAYER']
+                        in_players_1.append(tmp_in_player)
+                    elif subs_team1.loc[i, 'PLAYTYPE'] == 'OUT':
+                        tmp_out_player = subs_team1.loc[i, 'PLAYER']
+                        out_players_1.append(tmp_out_player)
+                
+                in_players_2 = []
```
Comment (giasemidis) [inline src/euroleague_api/team_lineup_euroleague_api.py:56]: I am confused, why do you capture the changes of team 2 here? These will be captured when a team 2 player will come "IN" later in the for loop. Why double counting?

File: src/euroleague_api/team_lineup_euroleague_api.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    # Filling start of the match with starters
+    on_court_mask1 = ['Team_A1', 'Team_A2', 'Team_A3', 'Team_A4', 'Team_A5']
+    on_court_mask2 = ['Team_B1', 'Team_B2', 'Team_B3', 'Team_B4', 'Team_B5']
+
+    pbp_data.loc[0, on_court_mask1] = starting_five.iloc[:, 0].tolist()
+    pbp_data.loc[0, on_court_mask2] = starting_five.iloc[:, 1].tolist()
+
+    # Capturing subs and replacing player columns with players who substitute in
+    players = pbp_data.loc[min(pbp_data.index.tolist()), on_court_cols]
```
Comment (giasemidis) [inline src/euroleague_api/team_lineup_euroleague_api.py]: if pbp_data.index is reset as suggested above, this line can be simplified as `players = pbp_data.loc[0, on_court_cols]`

File: src/euroleague_api/team_lineup_euroleague_api.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+                            players[j] = subs_dict[key]
+                        else:
+                            continue
+                
+                pbp_data.loc[max(all_subs.index), on_court_cols] = players # replace them in the actualy dataframe
+            
+            else:
+                pass
+    
+    pbp_data.loc[:, on_court_cols] = pbp_data.loc[:, on_court_cols].fillna(method='ffill') # filling every play
```
Comment (giasemidis) [inline src/euroleague_api/team_lineup_euroleague_api.py:86]: Since you are fillining the non-sub rows at the end, why don't you iterate and in the outer for loop only the in/our sub rows i.e. `sub_pbp_data = pbp_data["PLAYTYPE"].isin(["IN", "OUT"]`. It will be much faster.

**Review by giasemidis — CHANGES_REQUESTED** [review]
You iterate over all rows in the dataframe but then you fill some rows for both teams, making the code hard to follow. The code would be easier to read if for each row, only a single change is made. In your implementation, multiple ones are happening, which it should be avoided.

File: src/euroleague_api/team_lineup_euroleague_api.py (hunk 5 above)
Comment (YigitAsik) [reply src/euroleague_api/team_lineup_euroleague_api.py:56]: I don't go through every substitution row, that team 2 player may never come later.

The structure of the data is strange, not like NBAs where every substitution corresponds to a single row. Here, every substitution corresponds to 2 rows and they don't always appear the same way (e.g., first 'IN' row after 'OUT' row), there is no apparent pattern to it. So, I get the exact time where the substitution occurred and check if there is any more sub made at that point. Then I have them in a dataframe and go from there.

In that dataframe, it doesn't necessarily mean that the first 'IN' row's team is same as the first 'OUT' row's team. That might be why I do it like that.

Or, I didn't get what you mean which is more likely :) But I don't "loop" through every sub, I loop and when I see one I get all the subs made at that particular point. I do it this way because of the structure of the data.

Comment (giasemidis) [reply src/euroleague_api/team_lineup_euroleague_api.py:56]: Sure, no problem. I will approve the PR and might do some more testing my side too. When I publish the package, I will ask you to test it.
